core/src/context.rs

```diff
@@ -2,21 +2,120 @@ use error::HolochainError;
 use holochain_agent::Agent;
 use logger::Logger;
 use persister::Persister;
-use std::sync::{Arc, Mutex};
+use state::State;
+use std::sync::{Arc, Mutex, RwLock, RwLockReadGuard};
 
-/// Context holds those aspects of the outside world that a Holochain instance needs to operate
+/// Context holds those aspects of a living Holochain instance that components need to operate.
+/// This includes modules that are injected from the outside like logger and persister
+/// but also the state of the instance that gets injected before passing on the context
+/// to inner components/reducers.
 #[derive(Clone)]
 pub struct Context {
     pub agent: Agent,
     pub logger: Arc<Mutex<Logger>>,
     pub persister: Arc<Mutex<Persister>>,
+    state: Option<Arc<RwLock<State>>>,
 }
 
 impl Context {
+    pub fn new(
+        agent: Agent,
+        logger: Arc<Mutex<Logger>>,
+        persister: Arc<Mutex<Persister>>,
+    ) -> Context {
+        Context {
+            agent,
+            logger,
+            persister,
+            state: None,
+        }
+    }
     // helper function to make it easier to call the logger
     pub fn log(&self, msg: &str) -> Result<(), HolochainError> {
         let mut logger = self.logger.lock().or(Err(HolochainError::LoggingError))?;
         logger.log(msg.to_string());
         Ok(())
     }
+
+    pub fn set_state(&mut self, state: Arc<RwLock<State>>) {
+        self.state = Some(state);
```

Comment:
how do we stop uncoordinated processes clobbering each other's state?

Reply:
@lucksus if A and B do concurrent reads, then each compute a new state then both call `set_state`, what happens?

Reply:
The state gets only changed in `process_action()` through the main event loop. The whole point of this pattern is that you get a clear sequence of state changes. If we have two threads running an event loop on the same instance, we have violated this architecture and need to fix that.

Reply:
i'm asking about concurrent reads not writes

state = 1
-- concurrent  --
A reads state 1
B reads state 1
A calculates state + 1 = 2
B calculates state + 2 = 3
-- sequential --
A sets state to 2
B sets state to 3

final state is 3

shouldn't state be 4?

Reply:
After our conversation during stand-up, can we resolve this here for now, @thedavidmeister?